packages/passwordless-auth-construct/src/sign-up/construct.ts

```diff
@@ -0,0 +1,85 @@
+import { Construct } from 'constructs';
+import { CustomAuthTriggers } from '../types.js';
+import {
+  NodejsFunction,
+  NodejsFunctionProps,
+  OutputFormat,
+} from 'aws-cdk-lib/aws-lambda-nodejs';
+import * as path from 'path';
+import { fileURLToPath } from 'url';
+import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda';
+import * as apigateway from 'aws-cdk-lib/aws-apigateway';
+import { IUserPool } from 'aws-cdk-lib/aws-cognito';
+import { Effect, Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
+
+const filename = fileURLToPath(import.meta.url);
+const dirname = path.dirname(filename);
+
+/**
+ * Amplify SignUp Passwordless construct
+ */
+export class AmplifySignUpPasswordless extends Construct {
+  /**
+   * Create a new SignUp Passwordless construct
+   * verifyAuthChallengeResponse function is needed in order to attach policy
+   * userPool is needed to identity the resource and create the policy for verifyAuthChallengeResponse
+   */
+  constructor(
+    scope: Construct,
+    id: string,
+    verifyAuthChallengeResponse: NodejsFunction,
+    userPool: IUserPool
+  ) {
+    super(scope, id);
+
+    // default memory allocation for lambda functions
+    const defaultMemorySize = 128;
+
+    const commonOptions: NodejsFunctionProps = {
+      entry: path.join(dirname, '.', 'create_user_service.js'),
+      runtime: Runtime.NODEJS_20_X,
+      architecture: Architecture.ARM_64,
+      bundling: {
+        format: OutputFormat.ESM,
+      },
+    };
+
+    const createUserFunction = new NodejsFunction(scope, `CreateUser${id}`, {
+      handler: 'createUser',
+      ...commonOptions,
+      memorySize: defaultMemorySize,
+    });
+
+    const api = new apigateway.RestApi(this, `CreateUserApi${id}`, {
+      restApiName: 'Create User service',
+      description: 'This service creates users.',
+    });
+
+    const putCreateUserIntegration = new apigateway.LambdaIntegration(
+      createUserFunction,
+      {
+        requestTemplates: { 'application/json': '{ "statusCode": "200" }' },
+      }
+    );
+
+    api.root.addMethod('PUT', putCreateUserIntegration);
+
+    const createUserPolicy = new Policy(this, `CreateUserPolicy${id}`, {});
+    createUserFunction.role!.attachInlinePolicy(createUserPolicy);
+    const createUserPolicyStatement = new PolicyStatement();
+    createUserPolicyStatement.addActions('cognito-idp:AdminCreateUser');
+    createUserPolicyStatement.addResources(userPool.userPoolArn);
+    createUserPolicyStatement.effect = Effect.ALLOW;
+    createUserPolicy.addStatements(createUserPolicyStatement);
```

Comment:
Question: Is there an advantage to using the more imperative API compared to:

```ts
createUserFunction.addToRolePolicy(
  new PolicyStatement({...})
);
```

For example, see:
https://github.com/aws-amplify/amplify-backend/blob/ca2f170f2e43b67a25178293bee3ecbc3b6ac432/packages/passwordless-auth-construct/src/construct.ts#L94-L115

Reply:
But I updated it to make it more readable, thanks